compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableOpInterface.cpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/GlobalOptimization/Interfaces/HoistableOpInterface.h"
+
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtDialect.h"
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+template <typename OpTy>
+struct UnhoistableOpInterface
+    : public IREE::Util::HoistableOpInterface::ExternalModel<
+          UnhoistableOpInterface<OpTy>, OpTy> {
+  bool isHoistableOp(Operation *) const { return false; }
+  bool isHoistableLeafOp(Operation *) const { return false; }
+};
+
+template <typename OpTy>
+struct HoistableNonLeafOpInterface
+    : public IREE::Util::HoistableOpInterface::ExternalModel<
+          HoistableNonLeafOpInterface<OpTy>, OpTy> {
+  bool isHoistableLeafOp(Operation *) const { return false; }
+};
+
+// The default interface is always hoistable. This acts as an override
+// for other default hoistability checks as the interface is checked
+// first.
+template <typename OpTy>
+struct AlwaysHoistableOpInterface
+    : public IREE::Util::HoistableOpInterface::ExternalModel<
+          AlwaysHoistableOpInterface<OpTy>, OpTy> {};
+
+template <typename OpTy>
+struct HoistableFlowOpInterface
+    : public IREE::Util::HoistableOpInterface::ExternalModel<
+          HoistableFlowOpInterface<OpTy>, OpTy> {
+  bool isHoistableLeafOp(Operation *op) const {
+    // We could skip the dyn_cast here if we split streamable flow ops and/or
+    // metadata flow ops out from the rest of the TableGen Flow op definitions.
+    auto streamableOp = llvm::dyn_cast<IREE::Flow::StreamableOpInterface>(op);
+    // All non-streamable ops are fine to hoist. Only avoid hoisting metadata
+    // and transfer ops as a leaf.
+    if (!streamableOp)
+      return true;
+    return !streamableOp.isMetadata() && !streamableOp.isTransfer();
+  }
+};
```

Comment:
instead of this, which is spooky action at a distance, we should just register this interface on ops we want to be hoistable and return true. since the interface is in Util and we own the dialect it's best just to implement it concretely. this acts as documentation (a user looking at the flow td or generated help can see which ops are hoistable) and avoids any brittleness - the flow dialect and any pass expected to work on it should always work regardless of whether global optimization is even compiled into the compiler binary. linalg is fine here as we don't own that and it's only hoisted/handled during this phase.

Reply:
Makes sense. I was sad when I realized the dynamic registration based on the other interface wasn't going to work the way I wanted. Should I drop the `Flow_StreamableOpInterface` for now then too? I don't think I'll have a use for it anymore (unless I made it something like an OpTrait maybe that could work, still not sure, but still seems wrong).